presto-main/src/main/java/com/facebook/presto/type/MapType.java

```diff
@@ -69,6 +72,115 @@ public Type getValueType()
     }
 
     @Override
+    public boolean isComparable()
+    {
+        return valueType.isComparable();
+    }
+
+    @Override
+    public int hash(Block block, int position)
+    {
+        Slice mapSlice = getSlice(block, position);
+        Block mapBlock = readStructuralBlock(mapSlice);
+        Map<Integer, Integer> hashMap = new HashMap<>();
+
+        for (int i = 0; i < mapBlock.getPositionCount(); i += 2) {
+            //TODO fix collision, hashPosition could always return 0
+            hashMap.put(hashPosition(keyType, mapBlock, position), hashPosition(valueType, mapBlock, position + 1));
+        }
+        return Objects.hash(hashMap);
+    }
+
+    @Override
+    public boolean equalTo(Block leftBlock, int leftPosition, Block rightBlock, int rightPosition)
+    {
+        Slice leftSlice = getSlice(leftBlock, leftPosition);
+        Slice rightSlice = getSlice(rightBlock, rightPosition);
+        Block leftMapBlock = readStructuralBlock(leftSlice);
+        Block rightMapBlock = readStructuralBlock(rightSlice);
+
+        Map<KeyWrapper, Integer> wrappedLeftMap = new HashMap<>();
+        for (int position = 0; position < leftMapBlock.getPositionCount(); position += 2) {
+            wrappedLeftMap.put(new KeyWrapper(keyType, leftMapBlock, position), position + 1);
+        }
+
+        Map<KeyWrapper, Integer> wrappedRightMap = new HashMap<>();
+        for (int position = 0; position < rightMapBlock.getPositionCount(); position += 2) {
+            wrappedRightMap.put(new KeyWrapper(keyType, rightMapBlock, position), position + 1);
+        }
+
+        if (wrappedLeftMap.size() != wrappedRightMap.size()) {
```

Comment:
you can move this before the two loops, and just check `leftMapBlock.getPositionCount() != rightMapBlock.getPositionCount()` that will improve performance in cases where the maps are different sizes